app/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,228 @@
+package tech.relaycorp.letro.awala
+
+import android.content.Context
+import android.util.Log
+import androidx.annotation.RawRes
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+import tech.relaycorp.awaladroid.Awala
+import tech.relaycorp.awaladroid.GatewayBindingException
+import tech.relaycorp.awaladroid.GatewayClient
+import tech.relaycorp.awaladroid.endpoint.FirstPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.InvalidThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.PublicThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.ThirdPartyEndpoint
+import tech.relaycorp.awaladroid.messaging.OutgoingMessage
+import tech.relaycorp.letro.R
+import tech.relaycorp.letro.awala.message.Message
+import tech.relaycorp.letro.awala.message.MessageRecipient
+import tech.relaycorp.letro.awala.message.MessageType
+import tech.relaycorp.letro.ui.navigation.Route
+import tech.relaycorp.letro.utils.awala.loadNonNullFirstPartyEndpoint
+import tech.relaycorp.letro.utils.awala.loadNonNullThirdPartyEndpoint
+import javax.inject.Inject
+
+interface AwalaManager {
+    val messages: Flow<Message>
+    suspend fun sendMessage(
+        message: Message,
+        recipient: MessageRecipient,
+    )
+    suspend fun isAwalaInstalled(currentScreen: Route): Boolean
+}
+
+class AwalaManagerImpl @Inject constructor(
+    private val awalaRepository: AwalaRepository,
+    @ApplicationContext private val context: Context,
+): AwalaManager {
+
+    private val awalaScope = CoroutineScope(Dispatchers.IO)
+
+    private val _messages = MutableSharedFlow<Message>()
+    override val messages: Flow<Message>
+        get() = _messages
+
+    @Volatile
+    private var isAwalaSetUp = false
+    private var awalaSetupJob: Job? = null
+
+    @Volatile
+    private var isAwalaInstalledOnDevice: Boolean? = null
+
+    @Volatile
+    private var isReceivingMessages = false
+
+    init {
+        awalaSetupJob = awalaScope.launch {
+            Awala.setUp(context)
+            checkIfAwalaAppInstalled()
```

Comment:
I think `startReceivingMessages()` should be called here, after `checkIfAwalaAppInstalled()` -- I think it's very weird to call it inside `checkIfAwalaAppInstalled()`.

Reply:
I put it there (in `configureAwala()` method) in order to start receiving messages after we authorised the client to receive messages from server. 

I call it after `Gateway.bind()` on purpose, because in the block you commented there is a possibility that awala is not installed and we will not start to receive messages after the next app start.

Reply:
> I put it there (in configureAwala() method) in order to start receiving messages after we authorised the client to receive messages from server.

I think maybe there's two parts to my original comment:

1. `checkIfAwalaAppInstalled()` is too greedy. It doesn't just check if Awala is installed -- it does much more than that.
2. `receiveMessages()` does not require any third-party endpoint to be authorised -- it just needs the first-party endpoint to be registered. (I should've written this down, but remember I said in a call that the other developer was going to fix that but didn't get round to)

> I call it after Gateway.bind() on purpose, because in the block you commented there is a possibility that awala is not installed and we will not start to receive messages after the next app start.

Fair enough, that makes sense, but is there something we can do to keep that behaviour whilst addressing the two points above?

Reply:
Sounds reasonable. I think we can start subscribing to messages after we've registered the first party endpoint then (by the same way I've implemented it for send auth after importing third party endpoint). But I would leave configuring Awala there to make sure that Awala is configured as soon as we estabilish the connection with Gateway. What do you think?

Reply:
Makes sense 👌🏾